### PR TITLE
New arch/update public share for file

### PR DESCRIPTION
**Pull request by davigonz**
Implements #2516 
Needs https://github.com/owncloud/android-library/pull/247

### CODE REVIEW

**Comment by davigonz** [comment]
@jesmrec UI tests finished, please have a look at them and tell me whether are enough or we could include more

File: owncloudApp/src/androidTest/java/com/owncloud/android/shares/db/OCShareDaoTest.kt [hunk 1]
```diff
@@ -260,8 +260,8 @@ class OCShareDaoTest {
         )
 
         assertThat(textShares, notNullValue())
-        assertEquals(textShares.size, 2)
-        assertEquals(textShares.get(0).name, "Text 1 link")
+        assertEquals(textShares.size, 1)
+        assertEquals(textShares.get(0).name, "Text 2 link")
```
Comment (jcmontesmartos) [inline owncloudApp/src/androidTest/java/com/owncloud/android/shares/db/OCShareDaoTest.kt]: FYI: First value is the expected value and the second one is the actual value.

You don't have to change but take it into account for coming code

File: owncloudApp/src/main/java/com/owncloud/android/operations/UpdateSharePermissionsOperation.kt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        path = share.path
+
+        // Update remote share with password
+        val updateOp = UpdateRemoteShareOperation(share.remoteId)
+        updateOp.permissions = permissions
+        var result = updateOp.execute(client)
+
+        if (result.isSuccess) {
+            val getShareOp = GetRemoteShareOperation(share.remoteId)
+            result = getShareOp.execute(client)
```
Comment (jcmontesmartos) [inline owncloudApp/src/main/java/com/owncloud/android/operations/UpdateSharePermissionsOperation.kt]: reusing variables?? Please, don't do it.

        ....
        val updateOpResult = updateOp.execute(client)

        if (!updateOpResult.isSuccess) {
            return updateOpResult
         }
         val getShareOp = GetRemoteShareOperation(share.remoteId)
         val getShareOpResult = getShareOp.execute(client)
         if (!getShareOpResult.isSuccess) {
             return getShareOpResult
         }
        val remoteShare = result.data.shares[0]
        // TODO check permissions are being saved
        updateData(remoteShare)
        return getShareOpResult

File: owncloudApp/src/main/java/com/owncloud/android/operations/UpdateSharePermissionsOperation.kt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            }
+        }
+
+        return result
+    }
+
+    private fun updateData(share: RemoteShare) {
+        // Update DB with the response
+        share.path = path   // TODO - check if may be moved to UpdateRemoteShareOperation
+        share.isFolder = path.endsWith(FileUtils.PATH_SEPARATOR)
```
Comment (jcmontesmartos) [inline owncloudApp/src/main/java/com/owncloud/android/operations/UpdateSharePermissionsOperation.kt]: Why don't you modify RemoteShare class to generate isFolder attribute automatically?

     class RemoteShare {
         ...
         val isFolder: Boolean = path.endsWith(FileUtils.PATH_SEPARATOR)
             private set
         ...
     }

File: owncloudApp/src/main/java/com/owncloud/android/providers/FileContentProvider.kt [hunk 4]
```diff
@@ -487,7 +486,7 @@ class FileContentProvider(val appExecutors: AppExecutors = AppExecutors()) : Con
         return results
     }
 
-    private inner class DataBaseHelper internal constructor(context: Context) :
+    private inner class DataBaseHelper internal constructor(context: Context?) :
```
Comment (jcmontesmartos) [inline owncloudApp/src/main/java/com/owncloud/android/providers/FileContentProvider.kt]: Can it be nullable?

File: owncloudApp/src/main/java/com/owncloud/android/shares/db/OCShare.kt [hunk 5]
```diff
@@ -66,9 +71,34 @@ data class OCShare(
     val name: String?,
     @ColumnInfo(name = ProviderTableMeta.OCSHARES_URL)
     val shareLink: String?
-) {
+) : Parcelable {
```
Comment (jcmontesmartos) [inline owncloudApp/src/main/java/com/owncloud/android/shares/db/OCShare.kt]: Why don't you use @Parcelize?

https://kotlinlang.org/docs/tutorials/android-plugin.html

Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/shares/db/OCShare.kt]: Didn't know it, thanks for the suggestion!

File: owncloudApp/src/main/java/com/owncloud/android/operations/UpdateSharePermissionsOperation.kt (hunk 2 above)
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/operations/UpdateSharePermissionsOperation.kt]: Legacy code again, rewriting it...

File: owncloudApp/src/main/java/com/owncloud/android/operations/UpdateSharePermissionsOperation.kt (hunk 3 above)
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/operations/UpdateSharePermissionsOperation.kt]: Done

File: owncloudApp/src/main/java/com/owncloud/android/providers/FileContentProvider.kt (hunk 4 above)
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/providers/FileContentProvider.kt]: Yes, done

File: owncloudApp/src/main/java/com/owncloud/android/shares/db/OCShare.kt (hunk 5 above)
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/shares/db/OCShare.kt]: But `@Parcelable` is [still experimental](https://kotlinlang.org/docs/tutorials/android-plugin.html#parcelable-implementations-generator), I would not use it for production apps, too brave :grimacing:

Comment (jcmontesmartos) [reply owncloudApp/src/main/java/com/owncloud/android/shares/db/OCShare.kt]: I will be stable in the coming release...

BTW: 🐔  😄

**Comment by jesmrec** [comment]
Let's perform some checks:

- [X] Edit name
- [X] Set password
- [X] Folder permission -> edit to Download / View
- [X] Folder permission -> edit to Download / View / Upload
- [X] Folder permission -> edit to Upload Only (File Drop)
- [X] Remove Password 
- [X] Set Expiration date
- [X] Remove Expiration date 

Approved

**Pull request merged**